js/privatebin.js

```diff
@@ -924,6 +953,56 @@ jQuery.PrivateBin = (function($, RawDeflate) {
             };
         }
 
+        /**
+         * get PBKDF2 protected credentials for server to validate password
+         *
+         * @name   CryptTool.getCredentials
+         * @function
+         * @param  {string} key
+         * @param  {string} password
+         * @return {string} derived key
+         */
+        me.getCredentials = async function(key, password)
+        {
+            return btoa(
+                arraybufferToString(
+                    await deriveCredentials(key, password)
+                )
+            );
+        }
+
+        /**
+         * get HMAC of paste ID and PBKDF2 protected credentials for server to validate
+         *
+         * @name   CryptTool.getToken
+         * @function
+         * @param  {string} id
+         * @param  {string} key
+         * @param  {string} password
+         * @return {string} decrypted message, empty if decryption failed
+         */
+        me.getToken = async function(id, key, password)
+        {
+            return btoa(
+                arraybufferToString(
+                    await window.crypto.subtle.sign(
+                        {name: 'HMAC'},
```

Comment:
Ah oaky, so you already use HMAC after stretching the password. Okay.

Though do you really need to first export that key and then re-import it? This looks not nice…

Reply:
The reason for this was described and discussed in the issue: During the paste creation, the creator doesn't yet have the paste ID - it gets generated by the server when the paste is stored. Hence we have to let the server generate the full HMAC, just like the visitors later on.

I also under no circumstance want to use the same PBKDF2-mechanism as is used for the actual decryption - we would send the server the key needed to decrypt the paste in that case.

But you did give me an idea: One more piece of information that both the creator and visitor share is the domain name and path of the URL (I'd strip the protocol and the query parts, as those are the same both times). We could use that as the salt or use the key as salt and domain/path + (optional) password as the data to store.

Reply:
Like that salt idea. Don't know about your other explanation though: What. The server calculates the HMAC?? That makes no sense…

Reply:
This was all explained [in here](https://github.com/PrivateBin/PrivateBin/issues/245#issuecomment-502360026).

But I would now rather prefer your opinion reg. a server side only nonce and zero password/key exchange as outlined in the [third paragraph of my comment below](https://github.com/PrivateBin/PrivateBin/pull/447#issuecomment-506967453).